File: utils/parsecard.py
#!/usr/bin/env python

import sys

#    25     0x0B P          DATA               0x0F P LRC
# 000...000 1101 0 ddd0p ddd0p ddd0p ... ddd0p 1111 1 ccccc 00000 00000 ... 00000

# DATA - groups of 5-bits (3 data, pad, parity)
# ddd0p ddd0p ddd0p ... ddd0p

# DATA - groups of 3-bits, bits reversed
# ddd ddd ddd ... ddd

# DATA - keep last 26 bits
# dd ddd ddd ddd ddd ddd ddd ddd ddd

# DATA - first and last parity, 26 bits
# pd ddd ddd ddd ddd ddd ddd ddd ddp

# DATA - remove parity bits, 24 bits
# d ddd ddd ddd ddd ddd ddd ddd dd

# DATA - groups of 4 bits
# dddd dddd dddd dddd dddd dddd

# Facility# Card#
# dddd dddd dddd dddd dddd dddd

path = sys.argv[1]

fd = file(path, "r")
lines = fd.readlines()
fd.close()

header = lines[0]

ret = ""
for line in lines[1:]:
	line = line.strip()
	(tm, data, clock, present) = line.split(",")
	clock = int(clock)
	data = 1-int(data)
	if (clock == 0):
		ret += str(data)



leading = ret[0:25]
ret = ret[25:]

print ret

#print "leading", leading
#print "rest", ret
#print ""

# Break up the bits into 5-bit segments
segments = []
n = 0
while n < len(ret):
	segment = ret[n:n+5]
	segments.append(segment)
	n += 5

# Strip off the trailing groups of zero
while segments[-1] == "00000":
	del segments[-1]

# Remove the B segment
del segments[0]
# Remove the LRC
del segments[-1]
# Remove the F segment
del segments[-1]

# Transform the segments into 3-bit chunks, with the bits reversed
a = ""
for segment in segments:
	part = segment[0:3]
	a += part[2]+part[1]+part[0]

# Keep the last 26 bits and discard the rest
a = a[-26:]
# Discard first and last parity bits
a = a[1:-1]

print a, len(a)
n = 0
lst = []
while n < len(a):
	lst.append(a[n:n+4])
	n += 4

# Extract out the facility and card codes
facility = lst[0]+lst[1]
card = lst[2]+lst[3]+lst[4]+lst[5]

print int(facility, 2)
print int(card, 2)

File: utils/parsecard.c
#include <stdio.h>

#define EVENP(d0,d1,d2,d3,parity)   (((d0)+(d1)+(d2)+(d3)+(parity)) % 2 == 0)

char rawData[] = "111111111111111111111111100101111101111011110111101111011110111101001001010111101011101111001100101010111011110000001111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111";
int rawPos = 0;

#define BUFLEN 26
unsigned char buffer[BUFLEN];
int bufferPos = 0;

void store_data(int data) 
{
    buffer[bufferPos] = data;
    bufferPos = (bufferPos+1) % BUFLEN;
}

int read_data()
{
    char b = rawData[rawPos];
    if (b == 0) return -1;
    rawPos++;
    if (b == '1') return 0;
    return 1;
}

int convert_binary(int start, int bits)
{
    int value = 0;
    int mod = 1;
    int n;
    for (n = bits-1; n >= 0; n--) {
        if (buffer[(bufferPos+start+n) % BUFLEN] == 1) {
            value += mod;
        }
        mod = mod * 2;
    }
    return value;
}

int main(void)
{
    int data = 0, d0, d1, d2, d3, parity, n;
    int facility, card;

    /* Skip the first 25 bits of data (make sure they are zeros) */
    for (n = 0; n < 25; n++)
    {
        data = read_data();
        if (data != 0) {
            /* Error */
            printf("leading zeros expected\n");
        }
    }

    /* Read in groups of 5-bits */
    int start = 1;
    while (1) 
    {
        /* The bits are sent LSB first */
        d0 = read_data();
        if (d0 == -1) {
            /* End of data */
            break;
        }

        /* Read the rest of the data */
        d1 = read_data();
        d2 = read_data();
        d3 = read_data();
        parity = read_data();

        if (d1 == -1 || d2 == -1 || d3 == -1 || parity == -1) {
            printf("premature end of data\n");
            break;
        }

        /* Verify the parity bit (odd parity) */
        if (EVENP(d0,d1,d2,d3,parity)) {
            printf("parity failure on 5-bit chunk\n");
        }

        if (start) {
            /* The first segment should be 0xB == 1011B */
            if (! (d0 & d1 & !d2 & d3) ) {
                /* Invalid starting segment */
                printf("invalid start segment %d %d %d %d\n", d0, d1, d2, d3);
            }
        } else {
            /* Check for the ending segment 0xF == 1111B */
            if (d0 & d1 & d2 & d3) {
                /* End of data sequence, the LRC follows */
                d0 = read_data();
                d1 = read_data();
                d2 = read_data();
                d3 = read_data();
                parity = read_data();
                if (EVENP(d0,d1,d2,d3,parity)) {
                    printf("LRC parity failure\n");
                }
                /* The parity for LRC checks out, verify the actual LRC */

                /* Consume the remaining zeros */
                while (1) {
                    data = read_data();
                    if (data == -1) {
                        break;
                    }
                    if (data != 0) {
                        printf("trailing data must be zeros\n");
                    }
                }
                break;
            }

            if (d3 != 0) {
                printf("data segments must have zero padding\n");
            }

            /* Otherwise this is a chunk of card data. Store it in a circular
             * buffer of 26 bits, so when we are finished iterating we are
             * left with the last 26 bits of input data (rest is discarded) */
            printf("%d%d%d\n", d2, d1, d0);
            store_data(d2);
            store_data(d1);
            store_data(d0);
        }
        start = 0;
    }

    /* Check the data parity (first and last bit */

    /* Dump the data */
    for (n = 1; n < BUFLEN-1; n++) {
        printf("%d", buffer[(bufferPos+n) % BUFLEN]);
    }
    printf("\n");

    /* Extract the facility ID */
    facility = convert_binary(1, 8);
    printf("%d\n", facility);

    card = convert_binary(9, 16);
    printf("%d\n", card);

}

